File: go/pkg/scheduler/scheduler.go
package scheduler

import (
	"kclone-go/pkg/models"
	"kclone-go/pkg/state"
)

// ChooseNode selects the best node for a pod using bin-packing
func ChooseNode(nodes []*models.Node, pod *models.Pod) *models.Node {
	var eligible []*models.Node
	for _, node := range nodes {
		// Skip nodes that are not ready
		if !node.Ready {
			continue
		}
		// Skip tainted nodes (simplified - in real K8s, pods can have tolerations)
		if len(node.Taints) > 0 {
			continue
		}
		if node.Fits(pod.Spec.CPURequest, pod.Spec.MemRequest) {
			eligible = append(eligible, node)
		}
	}

	if len(eligible) == 0 {
		return nil
	}

	// Score nodes by available CPU and memory (higher score preferred)
	best := eligible[0]
	bestScore := best.CPUAvailable()*1000 + best.MemAvailable()
	for _, n := range eligible[1:] {
		score := n.CPUAvailable()*1000 + n.MemAvailable()
		if score > bestScore {
			best = n
			bestScore = score
		}
	}
	return best
}

// SchedulePendingPods attempts to schedule all pending pods
func SchedulePendingPods(cs *state.ClusterState) {
	pods := cs.GetPods()
	nodes := cs.GetNodes()

	for _, pod := range pods {
		if pod.Status.Phase == "Pending" {
			node := ChooseNode(nodes, pod)
			if node != nil {
				cs.BindPod(pod.UID, node.Name)
			} else {
				cs.SetPodFailed(pod.UID, "No nodes have enough free resources")
			}
		}
	}
}

File: src/kclone/scheduler.py
from __future__ import annotations

from typing import List, Optional

from .models import Node, Pod
from .state import ClusterState


def choose_node(nodes: List[Node], pod: Pod) -> Optional[Node]:
    eligible = [n for n in nodes if n.fits(pod.spec.cpu_request, pod.spec.mem_request)]
    if not eligible:
        return None
    eligible.sort(key=lambda n: (n.cpu_allocated + n.mem_allocated, -n.cpu_capacity - n.mem_capacity))
    return eligible[0]


def schedule_pending_pods(state: ClusterState) -> None:
    pending = [p for p in state.pods.values() if p.status.phase == "Pending"]
    for pod in pending:
        node = choose_node(list(state.nodes.values()), pod)
        if node:
            state.bind_pod(pod.uid, node.name)
        else:
            state.set_pod_failed(pod.uid, "No nodes have enough free resources")
